pkg/server/auth/auth.go

```diff
@@ -0,0 +1,130 @@
+package auth
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+)
+
+const (
+	// StateCookieName is the name of the cookie that holds state during auth flow.
+	StateCookieName = "state"
+	// IDTokenCookieName is the name of the cookie that holds the ID Token once
+	// the user has authenticated successfully with the OIDC Provider.
+	IDTokenCookieName = "id_token"
+	// RefreshTokenCookieName is the name of the cookie that holds the refresh
+	// token.
+	RefreshTokenCookieName = "refresh_token"
+	// ScopeProfile is the "profile" scope
+	ScopeProfile = "profile"
+	// ScopeEmail is the "email" scope
+	ScopeEmail = "email"
+)
+
+// RegisterAuthHandler registers the /callback route under a specified prefix.
+// This route is called by the OIDC Provider in order to pass back state after
+// the authentication flow completes.
+func RegisterAuthHandler(mux *http.ServeMux, prefix string, cfg *AuthConfig) {
+	mux.Handle(prefix+"/callback", cfg.callback())
+}
+
+type principalCtxKey struct{}
+
+// UserPrincipal is a simple model for the user, including their ID and Groups.
+type UserPrincipal struct {
+	ID     string   `json:"id"`
+	Groups []string `json:"groups"`
+}
+
+// WithPrincipal sets the principal into the context.
+func WithPrincipal(ctx context.Context, p *UserPrincipal) context.Context {
+	return context.WithValue(ctx, principalCtxKey{}, p)
+}
+
+// WithAPIAuth middleware adds auth validation to API handlers.
+//
+// Unauthorized requests will be denied with a 401 status code.
+func WithAPIAuth(next http.Handler, cfg *AuthConfig) http.Handler {
+	cookieAuth := NewJWTCookiePrincipalGetter(cfg.logger,
+		cfg.verifier(), IDTokenCookieName)
+	headerAuth := NewJWTAuthorizationHeaderPrincipalGetter(cfg.logger, cfg.verifier())
+
+	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		principal, err := MultiAuthPrincipal{cookieAuth, headerAuth}.Principal(r)
+		if err != nil || principal == nil {
+			cfg.logger.Error(err, "failed to get principal")
```

Comment:
err might be nil, because principal could be nil?

Also, you could create the `MultiAuthPrincipal` in the closure, rather than on each request?